sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -4034,3 +4009,35 @@ object ApplyCharTypePadding extends Rule[LogicalPlan] {
     if (targetLength > charLength) StringRPad(expr, Literal(targetLength)) else expr
   }
 }
+
+/**
+ * This rule removes metadata columns from `DataSourceV2Relation` under 2 cases:
+ *   - A single v2 scan (can be produced by `spark.table`), which is similar to star expansion, and
+ *     metadata columns should only be picked by explicit references.
+ *   - V2 scans under writing commands, as we can't insert into metadata columns.
+ */
+object RemoveMetadataColumns extends Rule[LogicalPlan] {
+  import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Implicits._
+
+  private def removeMetaCol(tbl: NamedRelation): NamedRelation = tbl match {
+    case r: DataSourceV2Relation =>
+      if (r.output.exists(_.isMetadataCol)) {
```

Comment:
Why to guard `r.copy(output = r.output.filterNot(_.isMetadataCol))`? Why not to do it always?